File: internal/core/materiality_test.go
// Fiscal convention: monetary values in the Drenyra ecosystem are int64 cents;
// no float is ever used for money. This test freezes the v0.4.0 MaterialityLevel
// contract: the declared classification is validated at write time, survives
// cloning, and does NOT participate in the envelope hash (frozen decision — the
// policy classifies by level; the hash bytes stay v0.3-identical).
package core

import (
	"strings"
	"testing"
)

func validMaterialityMemory() AccountingMemory {
	return AccountingMemory{
		Identity: Identity{ID: "mem-1", TopicKey: "dk/acme/202601/fact-1"},
		Title:    "material adjustment",
		Kind:     KindDecision,
		Scope: Scope{
			Kind:           ScopeKindCompany,
			OrganizationID: "tenant-1",
			CompanyID:      "acme",
			RUC:            "20100039201",
			Period:         "202601",
		},
		Content: Content{
			What:    "adjustment",
			Why:     "late document",
			Where:   "SUNAT",
			Learned: "pending review",
		},
		Status:       StatusPendingReview,
		FiscalEffect: FiscalEffectAdjustment,
		EffectiveAt:  "2026-01-15T00:00:00Z",
		RecordedAt:   "2026-01-16T00:00:00Z",
		Source: Source{
			System:    "drenyra-core",
			ActorID:   "agent-1",
			ActorKind: ActorKindAgent,
		},
		ContentHash: "content-hash",
	}
}

func TestIsValidMaterialityLevel(t *testing.T) {
	tests := []struct {
		level MaterialityLevel
		want  bool
	}{
		{MaterialityNormal, true},
		{MaterialityMaterial, true},
		{MaterialityCritical, true},
		{"", false},
		{"huge", false},
		{"Material", false},
	}
	for _, tt := range tests {
		t.Run(string(tt.level), func(t *testing.T) {
			if got := IsValidMaterialityLevel(tt.level); got != tt.want {
				t.Errorf("IsValidMaterialityLevel(%q) = %v, want %v", tt.level, got, tt.want)
			}
		})
	}
}

func TestAssertValidMemoryAcceptsKnownMaterialityLevels(t *testing.T) {
	for _, level := range []MaterialityLevel{MaterialityNormal, MaterialityMaterial, MaterialityCritical} {
		m := validMaterialityMemory()
		m.MaterialityLevel = &level
		if err := AssertValidMemory(m); err != nil {
			t.Errorf("AssertValidMemory rejected known level %q: %v", level, err)
		}
	}
	// NULL (nil pointer) stays valid and means normal.
	if err := AssertValidMemory(validMaterialityMemory()); err != nil {
		t.Errorf("AssertValidMemory rejected NULL level: %v", err)
	}
}

func TestAssertValidMemoryRejectsUnknownMaterialityLevel(t *testing.T) {
	bad := MaterialityLevel("mega")
	m := validMaterialityMemory()
	m.MaterialityLevel = &bad
	err := AssertValidMemory(m)
	if err == nil {
		t.Fatal("AssertValidMemory accepted unknown materiality level")
	}
	if !strings.Contains(err.Error(), "INVALID_MATERIALITY_LEVEL") {
		t.Errorf("error must carry INVALID_MATERIALITY_LEVEL, got %v", err)
	}
}

func TestCloneMemoryCarriesMaterialityLevel(t *testing.T) {
	level := MaterialityCritical
	m := validMaterialityMemory()
	m.MaterialityLevel = &level

	cloned := CloneMemory(m)
	if cloned.MaterialityLevel == nil || *cloned.MaterialityLevel != MaterialityCritical {
		t.Fatalf("clone lost materialityLevel: got %v, want critical", cloned.MaterialityLevel)
	}

	// The clone is defensive: mutating the source pointer must not affect it.
	normal := MaterialityNormal
	m.MaterialityLevel = &normal
	if *cloned.MaterialityLevel != MaterialityCritical {
		t.Errorf("clone aliases the source materialityLevel pointer")
	}
}

func TestEnvelopeHashIgnoresMaterialityLevel(t *testing.T) {
	base := validMaterialityMemory()
	base.ContentHash = ComputeContentHash(base)
	base.IdentityHash = ComputeIdentityHash(base)
	base.EnvelopeHash = ComputeEnvelopeHash(base)

	critical := MaterialityCritical
	raised := CloneMemory(base)
	raised.MaterialityLevel = &critical
	raised.EnvelopeHash = ComputeEnvelopeHash(raised)

	if raised.EnvelopeHash != base.EnvelopeHash {
		t.Errorf("envelope hash changed with materialityLevel: base %q, raised %q — MaterialityLevel must NOT participate (frozen decision)", base.EnvelopeHash, raised.EnvelopeHash)
	}
}

File: core/__tests__/materiality.test.ts
/**
 * Fiscal convention: monetary values in the Drenyra ecosystem are BigInt cents;
 * no float is ever used for money.
 *
 * MaterialityLevel mirror tests (v0.4.0 Step 1): validation accepts known
 * levels and rejects unknown ones, cloning carries the declared level, and the
 * envelope hash stays byte-identical when the level is set (frozen decision —
 * the policy classifies by level; the hash bytes remain v0.3-identical).
 */

import { describe, expect, it } from "vitest";

import type {
	AccountingMemory,
	MaterialityLevel,
} from "../types.js";
import {
	assertValidMemory,
	cloneMemory,
	computeContentHash,
	computeEnvelopeHash,
	computeIdentityHash,
	MATERIALITY_LEVELS,
} from "../types.js";

function validMemory(): AccountingMemory {
	return {
		identity: { id: "mem-1", topicKey: "dk/acme/202601/fact-1" },
		title: "material adjustment",
		kind: "decision",
		scope: {
			kind: "company",
			organizationId: "tenant-1",
			companyId: "acme",
			ruc: "20100039201",
			period: "202601",
		},
		content: {
			what: "adjustment",
			why: "late document",
			where: "SUNAT",
			learned: "pending review",
		},
		status: "pending_review",
		fiscalEffect: "adjustment",
		effectiveAt: "2026-01-15T00:00:00Z",
		recordedAt: "2026-01-16T00:00:00Z",
		source: {
			system: "drenyra-core",
			actorId: "agent-1",
			actorKind: "agent",
		},
		contentHash: "content-hash",
		revision: 1,
	};
}

describe("MaterialityLevel mirror", () => {
	it("exposes the frozen const set", () => {
		expect(MATERIALITY_LEVELS).toEqual(["normal", "material", "critical"]);
	});

	it.each(["normal", "material", "critical"] as const)(
		"assertValidMemory accepts level %s",
		(level: MaterialityLevel) => {
			const memory = validMemory();
			memory.materialityLevel = level;
			expect(() => assertValidMemory(memory)).not.toThrow();
		},
	);

	it("treats NULL (unset) as valid and normal", () => {
		expect(() => assertValidMemory(validMemory())).not.toThrow();
	});

	it("assertValidMemory rejects an unknown level", () => {
		const memory = validMemory();
		memory.materialityLevel = "mega" as MaterialityLevel;
		expect(() => assertValidMemory(memory)).toThrow(/INVALID_MATERIALITY_LEVEL/);
	});

	it("cloneMemory carries the declared level defensively", () => {
		const memory = validMemory();
		memory.materialityLevel = "critical";
		const cloned = cloneMemory(memory);
		expect(cloned.materialityLevel).toBe("critical");
		memory.materialityLevel = "normal";
		expect(cloned.materialityLevel).toBe("critical");
	});

	it("envelope hash is byte-identical when materialityLevel is set", async () => {
		const base = validMemory();
		base.contentHash = await computeContentHash(base);
		base.identityHash = await computeIdentityHash(base);
		const baseHash = await computeEnvelopeHash(base);

		const raised = cloneMemory(base);
		raised.materialityLevel = "critical";
		const raisedHash = await computeEnvelopeHash(raised);

		expect(raisedHash).toBe(baseHash);
	});
});
